rectools/visuals/metrics_app.py

```diff
@@ -85,62 +96,160 @@ def construct(
+        >>> metadata_df = pd.DataFrame(
+        ...    {
+        ...        Columns.Model: ["Model1", "Model2"],
+        ...        "factors": [64, 32],
+        ...        "regularization": [0.05, 0.05],
+        ...        "alpha": [2.0, 0.5],
+        ...    })
         >>> app = MetricsApp.construct(
-        ...    models_metrics=example_df,
+        ...    models_metrics=metrics_df,
+        ...    models_metadata=metadata_df,
         ...    show_legend=True,
         ...    auto_display=False,
-        ...    layout_kwargs={"width": 800, "height": 600})
+        ...    scatter_kwargs={"width": 800, "height": 600})
 
         Get plotly chart from the current widget state
 
         >>> fig = app.fig
         >>> fig = fig.update_layout(title="Metrics comparison")
         """
-        return cls(models_metrics, show_legend, auto_display, layout_kwargs)
+        cls._validate_models_metrics_base(models_metrics)
+        cls._validate_models_metrics_split(models_metrics)
+        cls._validate_models_metrics_names(models_metrics)
+        if models_metadata is None:
+            models_metadata = models_metrics[Columns.Model].drop_duplicates().to_frame()
+        cls._validate_models_metadata(models_metadata)
 
-    @property
-    @lru_cache
-    def metric_names(self) -> tp.List[str]:
-        """List of metric column names from the `models_metrics`."""
-        non_metric_columns = {Columns.Split, Columns.Model}
-        return [col for col in self.models_metrics.columns if col not in non_metric_columns]
+        merged_data = models_metrics.merge(models_metadata, on=Columns.Model, how="left")
+        merged_data[Columns.Model] = merged_data[Columns.Model].str.replace(" ", "_")
+
+        metric_names = [col for col in models_metrics.columns if col not in {Columns.Split, Columns.Model}]
+        meta_names = [col for col in models_metadata.columns if col != Columns.Model]
+
+        return cls(merged_data, metric_names, meta_names, show_legend, auto_display, scatter_kwargs)
 
     @property
     @lru_cache
     def model_names(self) -> tp.List[str]:
         """Sorted list of model names from `models_metrics`."""
-        return sorted(self.models_metrics[Columns.Model].unique())
+        return sorted(self.data[Columns.Model].unique())
 
     @property
-    def fold_ids(self) -> tp.List[int]:
+    @lru_cache
+    def fold_ids(self) -> tp.Optional[tp.List[int]]:
         """Sorted list of fold identifiers from the `models_metrics`."""
-        return sorted(self.models_metrics[Columns.Split].unique())
+        if Columns.Split in self.data.columns:
+            return sorted(self.data[Columns.Split].unique())
+        return None
 
-    def _validate_input_data(self) -> None:
-        if not isinstance(self.models_metrics, pd.DataFrame):
+    @staticmethod
+    def _validate_models_metrics_base(models_metrics: pd.DataFrame) -> None:
+        if not isinstance(models_metrics, pd.DataFrame):
```

Comment:
We can drop this check because we have type annotation that `models_metics` is a `pd.DataFrame`